gp/local_search/gradient_based.py

```diff
@@ -0,0 +1,56 @@
+try:
+    import torch
+except ModuleNotFoundError:
+    torch = None
+
+
+def gradient_based(individuals, objective, lr, gradient_steps, optimizer=None, clip_value=1e4):
+    """Perform a local search for numeric leaf values for the list of
+    individuals based on gradient information obtained via automatic
+    differentiation.
+
+    Parameters
+    ----------
+    individuals : List
+        List of individuals for which to perform local search.
+    objective : Callable
+        Objective function that is called with a differentiable graph
+        and returns a differentiable loss.
+    lr : float
+        Learning rate for optimizer.
+    gradient_steps : int
+        Number of gradient steps per individual.
+    optimizer : torch.optim.Optimizer, optional
+        Optimizer to use for parameter updates. Defaults to
+        torch.optim.SGD.
+    clip_value : float, optional
+        Clipping value for gradients. Clipping is skipped when set to
+        None. Defaults to 1e4.
+
+    """
+    if torch is None:
+        raise ModuleNotFoundError("No module named 'torch' (extra requirement)")
+
+    if optimizer is None:
+        optimizer_class = torch.optim.SGD
```

Comment:
i knew i should've commented on this ;) we can't do that bc we first need to check in line 31 whether torch is available. otherwise it would make sense as you suggest.

Reply:
sure, we could. i just handled it like this bc we do it similarly in other parts, e.g., the tests. i guess the question is whether this remains the only function in this module or not. also, i'm not sure when to raise the error if we put the check in the `__init__`? when the users tries to access `gradient_based`?

Reply:
I mean, if `torch` is not available, this module should not be available. If you do something like this in `__init__.py` this should work, right?
```python
try:
    import torch  # noqa: F401
    from .gradient_based import gradient_based
except ModuleNotFoundError:
    pass

Reply:
yes, of course. but how do we handle the case where torch is not available and the user tries to `local_search.gradient_based`. they would just get a noninformative "ModuleNotFoundError", right? the way it's currently implemented they will have the information that this exists, but they need the extra requirement to use it.

Reply:
Okay, I see your point. What about doing this then:
```python

try:
    import torch
except ModuleNotFoundError:
    torch = None

if torch is None:
    def gradient_based(*args, **kwargs):
          raise ModuleNotFoundError("No module named 'torch' (extra requirement)")
else:
    def gradient_based(..., optimizer=torch.optim.SGD, ...):
        ## full function definition
```

Reply:
interesting, haven't thought of that :thinking: do you think this is easier to understand than just accepting a `None` as a default? or are you afraid this is going to lead to problems with mypy? besides the seeemingly increased complexity, the additional indentation bugs me

Reply:
I think it is easier to understand, but I get the point with the additional indent. Maybe it's fine to leave it as it is.